sycl/include/CL/sycl/intel/sub_group.hpp

```diff
@@ -25,69 +31,157 @@ template <typename T, access::address_space Space> class multi_ptr;
+    using OCLTy = detail::ConvertToOpenCLType_t<T>;                            \
+    OCLTy Arg = detail::convertDataToType<T, OCLTy>(x);                        \
+    OCLTy Ret =                                                                \
+        __spirv_##SPIRVOperation(Arg, static_cast<uint32_t>(local_id.get(0))); \
+    return detail::convertDataToType<OCLTy, T>(Ret);                           \
+  }
 
-template <typename T, __spv::GroupOperation O>
-static typename std::enable_if<detail::is_floating_point<T>::value, T>::type
-calc(T x, intel::minimum<T> op) {
-  return __spirv_GroupFMin(__spv::Scope::Subgroup, O, x);
-}
+GENERATE_BODY_1ARG(shuffle, SubgroupShuffleINTEL)
+GENERATE_BODY_1ARG(shuffle_xor, SubgroupShuffleXorINTEL)
+
+#undef GENERATE_BODY_1ARG
+
+#define GENERATE_BODY_2ARG(name, SPIRVOperation)                               \
+  template <typename T> T name(T A, T B, uint32_t Delta) {                     \
+    using OCLTy = detail::ConvertToOpenCLType_t<T>;                            \
+    OCLTy Arg1 = detail::convertDataToType<T, OCLTy>(A);                       \
+    OCLTy Arg2 = detail::convertDataToType<T, OCLTy>(B);                       \
+    OCLTy Ret = __spirv_##SPIRVOperation(Arg1, Arg2, Delta);                   \
+    return detail::convertDataToType<OCLTy, T>(Ret);                           \
+  }
 
-template <typename T, __spv::GroupOperation O>
-static typename std::enable_if<
-    !detail::is_floating_point<T>::value && std::is_signed<T>::value, T>::type
-calc(T x, intel::maximum<T> op) {
-  return __spirv_GroupSMax(__spv::Scope::Subgroup, O, x);
+GENERATE_BODY_2ARG(shuffle_down, SubgroupShuffleDownINTEL)
+GENERATE_BODY_2ARG(shuffle_up, SubgroupShuffleUpINTEL)
+
+#undef GENERATE_BODY_2ARG
+
+// Selects 8-bit, 16-bit or 32-bit type depending on size of T. If T doesn't
+// maps to mentioned types, then void is returned
+template <typename T>
+using SelectBlockTy =
+    select_apply_cl_scalar_t<T, uint8_t, uint16_t, uint32_t, void>;
+
+template <typename T, access::address_space Space>
+using AcceptableForLoadStore =
+    bool_constant<!std::is_same<void, SelectBlockTy<T>>::value &&
+                  Space == access::address_space::global_space>;
+
+// TODO: move this to public cl::sycl::bit_cast as extension?
+template <typename To, typename From>
+To bit_cast(const From &from) {
+#if __cpp_lib_bit_cast
+  return std::bit_cast<To>(from);
+#else
+  To to;
+  std::memcpy(&to, &from, sizeof(To));
+  return to;
+#endif // __cpp_lib_bit_cast
 }
 
-template <typename T, __spv::GroupOperation O>
-static typename std::enable_if<
-    !detail::is_floating_point<T>::value && std::is_unsigned<T>::value, T>::type
-calc(T x, intel::maximum<T> op) {
-  return __spirv_GroupUMax(__spv::Scope::Subgroup, O, x);
+template <typename T, access::address_space Space>
+T load(const multi_ptr<T, Space> src) {
+  using Ty = SelectBlockTy<T>;
+  using GlobalPtrTy = __attribute__((opencl_global)) Ty *;
```

Comment:
TODO: use `detail::ConvertToOpenCLType_t` here to avoid explicit use of address space attribute

Reply:
Done